File: impl/scripts/aya-sev-snp/producer_vsock_smoke.exs
# Producer vsock smoke — Elixir edition (TASK-122 AC#3 staging evidence)
#
# Drives the 2d-hsm staging enclave through a vsock↔UDS relay (Python) using the
# REAL Chain.ProducerHsm.Wire codec (the 2D producer signer client's CBOR layer).
# This is the literal "block producer uses 2d-hsm" evidence.
#
# Setup on aya:
#   /opt/elixir-1.16/bin/elixirc wire.ex          # compile Wire module
#   enclave-vsock-staging &                        # staging enclave on CID=1:5000
#   python3 vsock_uds_relay.py /tmp/phsm.sock 1 5000 &  # vsock↔UDS relay
#   /opt/elixir-1.16/bin/elixir producer_vsock_smoke.exs

defmodule Smoke do
  alias Chain.ProducerHsm.Wire

  @proto_ver 1
  @uds_path System.get_env("PRODUCER_HSM_UDS", "/tmp/phsm.sock")

  # Connect via UDS to the vsock↔UDS relay
  def connect do
    opts = [:binary, active: false, packet: :raw]
    case :gen_tcp.connect({:local, String.to_charlist(@uds_path)}, 0, opts, 5000) do
      {:ok, s} -> s
      {:error, reason} -> raise "UDS connect to #{@uds_path} failed: #{inspect(reason)}"
    end
  end

  def send_frame(s, type_byte, payload) do
    total = 2 + byte_size(payload)
    frame = <<total::32-big, @proto_ver, type_byte, payload::binary>>
    :ok = :gen_tcp.send(s, frame)
    :ok = :inet.setopts(s, recbuf: 1_048_576)
  end

  def recv_frame(s) do
    {:ok, <<total::32-big>>} = :gen_tcp.recv(s, 4, 5000)
    {:ok, <<@proto_ver, type, payload::binary>>} = :gen_tcp.recv(s, total, 5000)
    {type, payload}
  end

  def round_trip(s, type_byte, payload) do
    send_frame(s, type_byte, payload)
    recv_frame(s)
  end

  def run do
    IO.puts("Producer Elixir smoke via UDS relay at #{@uds_path}")
    s = connect()
    IO.puts("Connected to enclave (through vsock↔UDS relay)")

    results = []

    # 1. GET_MEASUREMENT
    IO.puts("\n=== GET_MEASUREMENT (0x01) ===")
    {:ok, gm_payload} = Wire.encode_get_measurement_request(%{version: 1})
    {0x01, gm_resp} = round_trip(s, 0x01, gm_payload)
    {:ok, gm} = Wire.decode_get_measurement_response(gm_resp)
    IO.puts("  measurement: #{byte_size(gm.measurement)} bytes")
    IO.puts("  pq_pubkey: #{byte_size(gm.pq_pubkey)} bytes (expect 1952)")
    IO.puts("  pq_signing_ready: #{gm.pq_signing_ready}")
    IO.puts("  supported_ticket_types: #{inspect(gm.supported_ticket_types)}")
    results = [check("GET_MEASUREMENT: 1952-byte pubkey + signing ready",
                      byte_size(gm.pq_pubkey) == 1952 and gm.pq_signing_ready == true) | results]

    # 2. GET_STATUS
    IO.puts("\n=== GET_STATUS (0x30) ===")
    {:ok, gs_payload} = Wire.encode_get_status_request(%{version: 1})
    {0x30, gs_resp} = round_trip(s, 0x30, gs_payload)
    {:ok, gs} = Wire.decode_get_status_response(gs_resp)
    IO.puts("  armed: #{gs.armed}")
    results = [check("GET_STATUS: disarmed", gs.armed == false) | results]

    # 3. SIGN_AUTHORIZATION_TICKET (recovery — enclave rejects without arming)
    IO.puts("\n=== SIGN_AUTHORIZATION_TICKET (0x10) ===")
    ticket = %{
      ticket_type: 0,
      nonce: 1,
      context_hash: :binary.copy(<<0xAB>>, 32),
      activation_height: 1000,
      new_measurement: :binary.copy(<<0x55>>, 48),
      pq_pubkey: gm.pq_pubkey,
      fork_spec_hash: nil,
      new_header_version: nil
    }
    {:ok, sat_payload} = Wire.encode_sign_authorization_ticket_request(%{ticket: ticket})
    {0x10, sat_resp} = round_trip(s, 0x10, sat_payload)

    if Wire.wire_error?(sat_resp) do
      {:ok, code, reason} = Wire.decode_wire_error(sat_resp)
      IO.puts("  wire error (expected): code=#{code} reason=#{reason}")
      results = [check("SIGN_AUTH: wire-error code=2", code == 2) | results]
    else
      {:ok, sat} = Wire.decode_sign_authorization_ticket_response(sat_resp)
      IO.puts("  SUCCESS: signature=#{byte_size(sat.signature)} bytes!")
      results = [check("SIGN_AUTH: 3309-byte signature", byte_size(sat.signature) == 3309) | results]
    end

    # 4. ARM_FOR_PRODUCTION (bogus proof → refused)
    IO.puts("\n=== ARM_FOR_PRODUCTION (0x20) ===")
    req = %{
      authorized_state: %{
        pq_pubkey: gm.pq_pubkey,
        measurement: gm.measurement,
        activated_at_height: 99,
        source_ticket_hash: :binary.copy(<<0xCC>>, 32)
      },
      recent_chain_proof: %{
        finalized_height: 100,
        finalized_header_hash: :binary.copy(<<0xDD>>, 32),
        recovery_history_tail: [:binary.copy(<<0xEE>>, 32)],
        proof_data: <<0x01>>,
        signature_from_recent_producer: :binary.copy(<<0xAA>>, 64)
      }
    }
    {:ok, arm_payload} = Wire.encode_arm_for_production_request(req)
    {0x20, arm_resp} = round_trip(s, 0x20, arm_payload)
    {:ok, arm} = Wire.decode_arm_for_production_response(arm_resp)
    IO.puts("  status: #{arm.status}")
    IO.puts("  reason: #{inspect(arm.reason)}")
    results = [check("ARM: refused (bogus proof)", arm.status == :refused) | results]

    :gen_tcp.close(s)

    IO.puts("\n============================================================")
    passed = Enum.count(results, & &1)
    total = length(results)
    IO.puts("#{passed}/#{total} checks passed")
    IO.puts("Producer signer client (Chain.ProducerHsm.Wire) drove the real")
    IO.puts("2d-hsm staging enclave via AF_VSOCK — TASK-122 AC#3 staging evidence.")

    if passed < total, do: System.halt(1)
  end

  defp check(label, true) do
    IO.puts("  [PASS] #{label}")
    true
  end
  defp check(label, false) do
    IO.puts("  [FAIL] #{label}")
    false
  end
end

Smoke.run()

File: impl/scripts/aya-sev-snp/producer_vsock_smoke.py
#!/usr/bin/env python3
"""Producer vsock smoke — exercises all 4 producer commands against the
staging enclave over AF_VSOCK (TASK-122 AC#3 staging evidence).

Usage:
  # Start the enclave first (loopback CID=1 or guest CID=42):
  #   env TWOD_HSM_VSOCK_CID=1 TWOD_HSM_VSOCK_PORT=5000 enclave-vsock-staging &
  # Then:
  VSOCK_CID=1 TWOD_HSM_VSOCK_PORT=5000 python3 producer_vsock_smoke.py
"""
from __future__ import annotations

import os
import socket
import struct
import sys
import time

if any(a in ("-h", "--help") for a in sys.argv[1:]):
    print(__doc__)
    sys.exit(0)

try:
    import cbor2  # type: ignore
except ImportError:
    sys.exit("Requires cbor2: apt install python3-cbor2 or pip install cbor2")

CID = int(os.environ.get("VSOCK_CID", "1"))
PORT = int(os.environ.get("TWOD_HSM_VSOCK_PORT", "5000"))
TIMEOUT = float(os.environ.get("VSOCK_SMOKE_TIMEOUT", "10"))
FAILURES: list[str] = []


def vsock_connect(cid: int, port: int, timeout: float) -> socket.socket:
    sock = socket.socket(socket.AF_VSOCK, socket.SOCK_STREAM)
    sock.settimeout(timeout)
    sock.connect((cid, port))
    return sock


def encode_frame(msg_type: int, payload: bytes) -> bytes:
    total = 2 + len(payload)
    return struct.pack(">I", total) + bytes([1, msg_type]) + payload


def recv_frame(sock: socket.socket, timeout: float) -> tuple[int, bytes]:
    deadline = time.monotonic() + timeout
    header = recv_until(sock, 4, timeout)
    total = struct.unpack(">I", header)[0]
    if total > 1_048_576:
        raise ValueError(f"frame too large: {total}")
    body = recv_until(sock, total, max(deadline - time.monotonic(), 0.1))
    if len(body) < 2:
        raise ValueError("frame body too short")
    version, msg_type = body[0], body[1]
    if version != 1:
        raise ValueError(f"unexpected protocol version: {version}")
    return msg_type, body[2:]


def recv_until(sock: socket.socket, n: int, timeout: float) -> bytes:
    deadline = time.monotonic() + timeout
    buf = b""
    while len(buf) < n:
        remaining = deadline - time.monotonic()
        if remaining <= 0:
            raise socket.timeout(f"recv {n} bytes: timed out (got {len(buf)})")
        sock.settimeout(remaining)
        chunk = sock.recv(min(8192, n - len(buf)))
        if not chunk:
            raise ConnectionError(f"peer closed (got {len(buf)}/{n})")
        buf += chunk
    return buf


def send_recv(msg_type: int, payload: bytes, label: str) -> tuple[int, bytes]:
    frame = encode_frame(msg_type, payload)
    sock = vsock_connect(CID, PORT, TIMEOUT)
    try:
        sock.sendall(frame)
        mt, resp_payload = recv_frame(sock, TIMEOUT)
        return mt, resp_payload
    finally:
        sock.close()


def check(cond: bool, msg: str) -> None:
    status = "OK" if cond else "FAIL"
    print(f"  [{status}] {msg}")
    if not cond:
        FAILURES.append(msg)


# Message type bytes (spec §8)
GET_MEASUREMENT = 0x01
SIGN_AUTHORIZATION_TICKET = 0x10
ARM_FOR_PRODUCTION = 0x20
GET_STATUS = 0x30


def test_get_measurement() -> None:
    print("\n=== GET_MEASUREMENT (0x01) ===")
    payload = cbor2.dumps({1: 1})  # version=1
    mt, resp = send_recv(GET_MEASUREMENT, payload, "GET_MEASUREMENT")
    check(mt == GET_MEASUREMENT, f"response type=0x{mt:02x} (expected 0x01)")
    m = cbor2.loads(resp)
    check(isinstance(m, dict), "response is CBOR map")
    check(m.get(1) == 1, f"version=1 (got {m.get(1)})")
    pubkey = m.get(4, b"")
    check(len(pubkey) == 1952, f"pq_pubkey is 1952 bytes ML-DSA-65 (got {len(pubkey)})")
    global ENCLAVE_PUBKEY
    ENCLAVE_PUBKEY = pubkey
    ready = m.get(6)
    check(ready is True, f"pq_signing_ready=True (got {ready})")
    types = m.get(5, [])
    check(types == [0, 1], f"supported_ticket_types=[0,1] (got {types})")
    print(f"  measurement: {len(m.get(2, b''))} bytes, attestation: {len(m.get(3, b''))} bytes")


def test_get_status() -> None:
    print("\n=== GET_STATUS (0x30) ===")
    payload = cbor2.dumps({1: 1})
    mt, resp = send_recv(GET_STATUS, payload, "GET_STATUS")
    check(mt == GET_STATUS, f"response type=0x{mt:02x} (expected 0x30)")
    m = cbor2.loads(resp)
    check(isinstance(m, dict), "response is CBOR map")
    check(m.get(1) == 1, f"version=1 (got {m.get(1)})")
    armed = m.get(2)
    check(armed is False, f"armed=False (not armed yet; got {armed})")
    print(f"  armed={armed}, last_known_block={m.get(9)}")


ENCLAVE_PUBKEY = None  # populated by test_get_measurement; used by test_sign_authorization_ticket


def test_sign_authorization_ticket() -> None:
    print("\n=== SIGN_AUTHORIZATION_TICKET (0x10) ===")
    # Use the REAL enclave pubkey from GET_MEASUREMENT (roborev 10811: using a
    # placeholder pubkey but accepting a signature could bless a mismatched-key bug).
    pubkey = ENCLAVE_PUBKEY or b"\x42" * 1952
    ticket = {
        1: 0,  # ticket_type = PRODUCER_RECOVERY
        2: 1,  # nonce
        3: b"\xAB" * 32,  # context_hash
        4: 1000,  # activation_height
        5: b"\x55" * 48,  # new_measurement
        6: pubkey,  # pq_pubkey (real enclave key from GET_MEASUREMENT)
        7: None,  # fork_spec_hash (null for recovery)
        8: None,  # new_header_version (null for recovery)
        9: None,  # governance_ref (always null in v1)
    }
    payload = cbor2.dumps({1: 1, 2: ticket})
    mt, resp = send_recv(SIGN_AUTHORIZATION_TICKET, payload, "SIGN_AUTHORIZATION_TICKET")
    check(mt == SIGN_AUTHORIZATION_TICKET, f"response type=0x{mt:02x} (expected 0x10)")
    m = cbor2.loads(resp)

    if isinstance(m.get(1), int) and isinstance(m.get(2), str):
        # Wire error shape {1: code, 2: reason}
        code = m[1]
        reason = m[2]
        check(code == 2, f"wire error code=2 (semantic error; got code={code})")
        print(f"  wire error (expected — enclave rejects unsigned ticket): code={code} reason={reason}")
    elif m.get(1) == 1 and isinstance(m.get(2), bytes):
        # Success: {1: version, 2: signature, 3: ticket_hash}
        sig = m[2]
        check(len(sig) == 3309, f"signature is 3309 bytes ML-DSA-65 (got {len(sig)})")
        ticket_hash = m.get(3, b"")
        check(len(ticket_hash) == 32, f"ticket_hash is 32 bytes (got {len(ticket_hash)})")
        print(f"  SUCCESS: signature={len(sig)} bytes, ticket_hash={ticket_hash.hex()[:16]}...")
    else:
        check(False, f"unexpected response shape: keys={list(m.keys())}")


def test_arm_for_production() -> None:
    print("\n=== ARM_FOR_PRODUCTION (0x20) ===")
    # ARM requires a valid RecentChainProof with a real Ed25519 signature.
    # Without one, the enclave should refuse (code=2 semantic error).
    authorized_state = {
        1: b"\x42" * 1952,  # pq_pubkey
        2: b"\x5A" * 48,  # measurement
        3: 99,  # activated_at_height
        4: b"\xCC" * 32,  # source_ticket_hash
    }
    recent_chain_proof = {
        1: 100,  # finalized_height
        2: b"\xDD" * 32,  # finalized_header_hash
        3: [b"\xEE" * 32],  # recovery_history_tail
        4: b"\x01",  # proof_data (format 0x01)
        5: b"\xAA" * 64,  # signature_from_recent_producer (bogus Ed25519)
    }
    payload = cbor2.dumps({1: 1, 2: authorized_state, 3: recent_chain_proof})
    mt, resp = send_recv(ARM_FOR_PRODUCTION, payload, "ARM_FOR_PRODUCTION")
    check(mt == ARM_FOR_PRODUCTION, f"response type=0x{mt:02x} (expected 0x20)")
    m = cbor2.loads(resp)
    if m.get(1) == "armed":
        check(False, "ARM should NOT succeed with a bogus RecentChainProof (security regression)")
    elif isinstance(m.get(1), int) and isinstance(m.get(2), str):
        code = m[1]
        reason = m[2]
        check(code == 2, f"refused with code=2 (expected — bogus proof; got code={code})")
        print(f"  refused (expected — bogus RecentChainProof): code={code} reason={reason}")
    else:
        check(False, f"unexpected response shape: keys={list(m.keys())}")


def main() -> None:
    print(f"Producer vsock smoke: CID={CID} PORT={PORT} TIMEOUT={TIMEOUT}s")
    test_get_measurement()
    test_get_status()
    test_sign_authorization_ticket()
    test_arm_for_production()

    print(f"\n{'='*60}")
    if FAILURES:
        print(f"FAILED: {len(FAILURES)} check(s)")
        for f in FAILURES:
            print(f"  - {f}")
        sys.exit(1)
    else:
        print("ALL PRODUCER COMMANDS PASSED — enclave handles the full vsock producer protocol")
        sys.exit(0)


if __name__ == "__main__":
    main()
